website/www/site/content/en/blog/apache-beam-flink-and-kubernetes.md

```diff
@@ -0,0 +1,355 @@
+
+<img class="center-block"
+src="/images/blog/apache-beam-flink-and-kubernetes/watermark-metrics.png"
+alt="Apache Beam Watermark Metrics">
+
+
+## Flink Open Source UI
+
+We use and extend Apache Flink Dashboard UI so that we can closely monitor jobs and tasks, such as the  checkpoint duration, size, and failure. One important extension we did is a job history page so that we can look back at a job's start/update timeline and details so that we can debug issues.
+
+<img class="center-block"
+src="/images/blog/apache-beam-flink-and-kubernetes/flink-checkpoint-ui.png"
+alt="Flink Checkpoint UI">
+
+
+
+## Dashboards and Alerting for Backlog and Latency
+
+We have ~30k jobs, and we want to closely monitor and alert those jobs which run into abnormal states, so that we can quickly intervene. We have created dashboards for each application so that we can show the top latency list, and alert them based on thresholds. This is one example of the timestamp based latency dashboard for one application. We can set the alerting if the latency is larger than a threshold, such as 10 minutes, for a certain time continuously:
+
+
+<img class="center-block"
+src="/images/blog/apache-beam-flink-and-kubernetes/latency-graph.png"
+alt="Latency Graph">
+
+
+Another example of backlogs based dashboards:
+
+
+<img class="center-block"
+src="/images/blog/apache-beam-flink-and-kubernetes/backlog-graph.png"
+alt="Backlog Graph">
+
+
+Alerting is threshold based with frequent metrics checking. If a threshold is met, and continues for a certain amount of times, we will alert our internal Slack channels, or PagerDuty for immediate attention. We tune the alerting so that the accuracy is high.
+
+
+# Cost Optimization: Strategies and Tuning
+
+A major goal for us to move to a self managed streaming service is cost efficiency. We have achieved the goal by saving more than half of the cost by several minor tunings. We can see more room for improvements further.
+
+These are just a few tips in our journey so far:
+
+
+
+1. Using GCS as Checkpointing storage.
+2. Reducing the write frequency to GCS
+3. Freedom to use different machine types. For example, in GCP N2D machines are 15% cheaper than N2 machines.
+4. Autoscaling tasks to use optimal resources while maintaining latency SLA (we will cover the details in Part 2)
+
+We describe more details for #1 and #2 related to checkpointing optimization as follows.
+
+**Google Cloud Storage and Checkpointing**
+
+GCS serves as our checkpoint store due to its cost-effectiveness, scalability, and durability. When working with GCS, several design considerations and best practices can optimize scaling and performance:
```

Comment:
```suggestion
We use Google Cloud Storage as our checkpoint store because it is cost-effective, scalable, and durable. When working with Google Cloud Storage, the following design considerations and best practices can help you optimize scaling and performance:
```